torch/csrc/autograd/variable.h

```diff
@@ -159,171 +264,135 @@ struct VariableViewImpl : public VariableImpl {
-  // _grad_fn field is stale if attr_version != version_counter.current_version()
-  int attr_version;
+  // grad_fn field is stale if attr_version !=
+  // version_counter.current_version()
+  uint32_t attr_version = 0;
 };
 
-inline Variable make_variable(at::Tensor data, bool requires_grad=false) {
-  if (!data.defined()) {
-    return Variable();
-  }
+//===----------------------------------------------------------------------===//
+//                        Variable Implementation
+//===----------------------------------------------------------------------===//
 
-#ifndef WITH_SCALARS
-  if (data.dim() == 0) {
-    // don't expose 0-dim tensors to Variable API.
-    data = data.as_strided_({1}, {1});
-  }
-#endif
+inline Variable::Variable(Variable::Impl* self, bool retain)
+    : at::Tensor(self, retain) {}
 
-  return Variable(new VariableImpl(std::move(data), requires_grad), false);
+inline const std::shared_ptr<Function>& Variable::grad_fn() const {
+  return get()->get_grad_fn();
 }
 
-inline Variable make_variable(at::Tensor data, int output_nr, std::shared_ptr<Function> grad_fn) {
-  if (!data.defined()) {
-    return Variable();
-  }
-
-#ifndef WITH_SCALARS
-  if (data.dim() == 0) {
-    // don't expose 0-dim tensors to Variable API.
-    data = data.as_strided_({1}, {1});
-  }
-#endif
+inline void Variable::set_grad_accumulator(
+    std::weak_ptr<Function> grad_accumulator) {
+  get()->grad_accumulator = std::move(grad_accumulator);
+}
 
-  return Variable(new VariableImpl(std::move(data), false, output_nr, std::move(grad_fn)), false);
+inline std::shared_ptr<Function> Variable::try_get_grad_accumulator() const {
+  return get()->grad_accumulator.lock();
 }
 
-Variable make_variable(at::Tensor data, std::shared_ptr<Function> grad_fn);
+inline std::shared_ptr<Function> Variable::grad_accumulator() const {
+  return get()->get_grad_accumulator();
+}
 
-inline Variable make_variable_view(Variable base, at::Tensor data, int output_nr=0,
-                                   std::shared_ptr<Function> grad_fn=nullptr) {
-  if (!data.defined()) {
-    return Variable();
-  }
+inline void Variable::set_gradient_edge(Edge&& edge) {
+  get()->grad_fn = std::move(edge.function);
+  get()->output_nr = edge.input_nr;
+}
 
-#ifndef WITH_SCALARS
-  if (data.dim() == 0) {
-    // don't expose 0-dim tensors to Variable API.
-    data = data.as_strided_({1}, {1});
-  }
-#endif
+inline int Variable::output_nr() const noexcept {
+  return get()->output_nr;
+}
 
-  return Variable(new VariableViewImpl(std::move(base), std::move(data), output_nr, std::move(grad_fn)), false);
+inline void Variable::set_requires_grad(bool requires_grad) {
+  get()->requires_grad = requires_grad;
 }
 
+inline void Variable::set_pyobj(PyObject* pyobj) {
+  get()->pyobj = pyobj;
+}
 
-inline Variable::Variable(VariableImpl * self, bool retain) : Tensor(self, retain) {
+inline PyObject* Variable::pyobj() const noexcept {
```

Comment:
Is there a reason why you only mark so few things `noexcept`? E.g. `set_pyobj` can't throw either

Reply:
I do it rather mechanically for getters since they rarely throw. I'll take another look at other methods, thanks.